### PR TITLE
Respect `#sourceLocation` directives in `SourceLocationConverter`

### PR DESCRIPTION
**Pull request by ahoppen**
Add a `presumedFile` and `presumedLine` property to `SourceLocation` that contains the file and line of the location while taking `#sourceLocation` directives into account.

The terms “presumed file” and “presumed line” have been taken from LLVM and the Swift compiler.

rdar://99187174

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 1]
```diff
@@ -13,19 +13,39 @@
...
   /// The file in which this location resides.
+  ///
+  /// ### See also
+  /// ``SourceLocation/presumedFile``
   public let file: String
 
+  /// The line of this location when respecting `#sourceLocation` directives.
+  ///
+  /// If the location hasn’t been adjusted using `#sourceLocation` directives,
+  /// this is the same as `file`.
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:38]:
```suggestion
  /// this is the same as `line`.
```

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 2]
```diff
@@ -95,6 +146,15 @@ public final class SourceLocationConverter {
...
   let endOfFile: AbsolutePosition
 
+  /// The information from all `#sourceLocation` directives in the file
+  /// necessary to compute presumed locations.
+  ///
+  /// - `sourceLine` is the line at which the `#sourceLocation` statement occurs
+  ///   within the current file.
+  /// - `fileArgument` is the `file` argument of the `#sourceLocation` directive.
+  /// - `lineArgument` is the `line` argument of the `#sourceLocation` directive.
+  var sourceLocationDirectives: [(sourceLine: Int, arguments: (file: String, line: Int)?)] = []
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:156]: Could you add a struct for this? Named tuples are okay but a nested named tuple is pushing it IMO 😅

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 3]
```diff
@@ -104,11 +164,31 @@ public final class SourceLocationConverter {
     self.source = tree.syntaxTextBytes
     (self.lines, endOfFile) = computeLines(tree: Syntax(tree))
     precondition(tree.byteSize == endOfFile.utf8Offset)
+
+    for directive in SourceLocationCollector.collectSourceLocations(in: tree) {
+      let location = self.physicalLocation(for: directive.positionAfterSkippingLeadingTrivia)
+      if let args = directive.args {
+        let logicalFile = args.fileName.segments.description
+        guard let logicalLine = Int(args.lineNumber.description) else {
+          // FIXME: Handle hex line numbers
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:173]: I'm not convinced we care about this case.

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 4]
```diff
@@ -104,11 +164,31 @@ public final class SourceLocationConverter {
     self.source = tree.syntaxTextBytes
     (self.lines, endOfFile) = computeLines(tree: Syntax(tree))
     precondition(tree.byteSize == endOfFile.utf8Offset)
+
+    for directive in SourceLocationCollector.collectSourceLocations(in: tree) {
+      let location = self.physicalLocation(for: directive.positionAfterSkippingLeadingTrivia)
+      if let args = directive.args {
+        let logicalFile = args.fileName.segments.description
+        guard let logicalLine = Int(args.lineNumber.description) else {
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:172]: Both `presumed` rather than `logical` to keep with the name used in the API?

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 5]
```diff
@@ -145,13 +225,40 @@ public final class SourceLocationConverter {
...
+    return physicalLocation
+  }
+
+  /// Compute the location of `position` without taking `#sourceLocation`
+  /// directives into account.
+  ///
+  /// If the position is
   /// exceeding the file length then the ``SourceLocation`` for the end of file
   /// is returned. If position is negative the location for start of file is
   /// returned.
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:258]: Why do we allow negative `AbsolutePosition`?

File: Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift [hunk 6]
```diff
@@ -12,6 +12,26 @@
 
 import XCTest
 @_spi(RawSyntax) import SwiftSyntax
+import SwiftSyntaxBuilder
+
+fileprivate func assertLogicalSourceLocation(
```
Comment (bnbarham) [inline Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift:17]: `assertPresumedLocation`

File: Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift [hunk 7]
```diff
@@ -43,4 +63,87 @@ final class SourceLocationConverterTests: XCTestCase {
...
+
+      #sourceLocation()
+
+      let a = 2
+      """,
+      presumedFile: "input.swift",
+      presumedLine: 5
+    )
+  }
 }
```
Comment (bnbarham) [inline Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift:149]: Could you add some tests for negative + hex + other failure cases (eg. string)?

File: Sources/SwiftSyntax/SourceLocation.swift [hunk 8]
```diff
@@ -47,11 +67,26 @@ public struct SourceLocation: Hashable, Codable, CustomDebugStringConvertible {
...
+  ///   - presumedFile: If the location has been adjusted using `#sourceLocation`
+  ///                   directives, the adjusted file. If `nil`, this defaults to
+  ///                   `file`.
+  public init(
+    line: Int,
+    column: Int,
+    offset: Int,
+    file: String,
+    presumedLine: Int? = nil,
+    presumedFile: String? = nil
```
Comment (bnbarham) [inline Sources/SwiftSyntax/SourceLocation.swift:82]: Add to debug description?

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 5 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SourceLocation.swift:258]: I don’t think there’s a real use case. Would you prefer to `assert` or `precondition`? 🤔

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 8 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SourceLocation.swift:82]: I am not super happy with the debug description as-is anyway because it also doesn’t include the file name and have been tempted to just remove it. WDYT?

File: Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift (hunk 7 above)
Comment (ahoppen) [reply Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift:149]: I added test case for hex, identifier and string interpolation inside the file name. Everything else should be disallowed by the syntax tree structure.

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SourceLocation.swift:173]: I added a doc comment saying that we ignore invalid `#sourceLocation` literals.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 8 above)
Comment (bnbarham) [reply Sources/SwiftSyntax/SourceLocation.swift:82]: Or just extend it to include file as well 🤷

File: Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift (hunk 7 above)
Comment (bnbarham) [reply Tests/SwiftSyntaxTest/SourceLocationConverterTests.swift:223]: Thanks!

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 8 above)
Comment (ahoppen) [reply Sources/SwiftSyntax/SourceLocation.swift:75]: Decided to remove `debugDescription`

> The debug description was lossy because it didn’t print file and offset. If we include that information, it’s so verbose that the debug description doesn’t really provide any benefit over the default debug description. Let’s just remove it.

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftSyntax/SourceLocation.swift (hunk 5 above)
Comment (bnbarham) [reply Sources/SwiftSyntax/SourceLocation.swift:285]: `precondition` seems fine to me

**Review by bnbarham — APPROVED** [review]

**Comment by ahoppen** [comment]
@swift-ci Please test Windows
